File: ch23p814textLayoutGeometry/ch23p811selfSizingTextField/MyLayoutManager.m


#import "MyLayoutManager.h"

@interface MyLayoutManager ()
@end

@implementation MyLayoutManager

-(void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origin {
    [super drawBackgroundForGlyphRange:glyphsToShow atPoint:origin];
    if (!self.wordRange.length)
        return;
    NSRange range = [self glyphRangeForCharacterRange:self.wordRange actualCharacterRange:nil];
    range = NSIntersectionRange(glyphsToShow, range);
    if (!range.length)
        return;
        
    CGRect r = [self boundingRectForGlyphRange:range inTextContainer:self.textContainers[0]];
    r.origin.x += origin.x;
    r.origin.y += origin.y;
    CGContextStrokeRect(UIGraphicsGetCurrentContext(), r);
}

@end

File: bk2ch10p543drawingWithTextKit/ch23p815attributedStringDrawing3/MyLayoutManager.swift


import UIKit

class MyLayoutManager : NSLayoutManager {
    
    var wordRange : NSRange = NSMakeRange(0,0)
    
    override func drawBackground(forGlyphRange glyphsToShow: NSRange, at origin: CGPoint) {
        super.drawBackground(forGlyphRange:glyphsToShow, at:origin)
        if self.wordRange.length == 0 {
            return
        }
        var range = self.glyphRange(forCharacterRange:self.wordRange, actualCharacterRange:nil)
        range = NSIntersectionRange(glyphsToShow, range)
        if range.length == 0 {
            return
        }
        if let tc = self.textContainer(forGlyphAt:range.location, effectiveRange:nil) {
            var r = self.boundingRect(forGlyphRange:range, in:tc)
            r.origin.x += origin.x
            r.origin.y += origin.y
            r = r.insetBy(dx: -2, dy: 0)
            let c = UIGraphicsGetCurrentContext()!
            c.saveGState()
            c.setStrokeColor(UIColor.black.cgColor)
            c.setLineWidth(1.0)
            c.stroke(r)
            c.restoreGState()
        }

    }
    
    
}
